File: services/emailService.ts
/**
 * Email Service - SendGrid Integration
 * Handles all email notifications for the workshop management system
 */

export interface EmailOptions {
  to: string;
  subject: string;
  text?: string;
  html?: string;
  templateId?: string;
  dynamicTemplateData?: Record<string, any>;
  attachments?: EmailAttachment[];
}

export interface EmailAttachment {
  content: string;  // Base64
  filename: string;
  type: string;
}

export interface EmailResult {
  success: boolean;
  messageId?: string;
  error?: string;
}

interface CompanyInfo {
  name: string;
  email: string;
  phone: string;
  address: string;
}

class EmailService {
  private serverUrl: string;
  private isConfigured: boolean = false;
  private company: CompanyInfo = {
    name: 'My Workshop',
    email: 'info@workshop.co.za',
    phone: '011 555 0000',
    address: '',
  };

  constructor() {
    // Get email server URL from environment
    this.serverUrl = import.meta.env.VITE_EMAIL_SERVER_URL || 'http://localhost:3001';
    this.isConfigured = !!this.serverUrl;
  }

  /**
   * Update company info used in email headers and signatures.
   * Call this once after loading the CompanyProfile from Firestore.
   */
  setCompanyInfo(info: Partial<CompanyInfo>): void {
    this.company = { ...this.company, ...info };
  }

  private getEmailHeader(): string {
    return `<h2 style="color: #2563eb;">${this.company.name}</h2>`;
  }

  private getEmailSignature(): string {
    const lines = [
      `Best regards,`,
      `<strong>${this.company.name}</strong>`,
      this.company.phone ? `Tel: ${this.company.phone}` : '',
      this.company.email ? `Email: <a href="mailto:${this.company.email}">${this.company.email}</a>` : '',
      this.company.address ? this.company.address : '',
    ].filter(Boolean).join('<br>');
    return `<p style="margin-top: 24px; border-top: 1px solid #e5e7eb; padding-top: 16px; color: #6b7280; font-size: 0.875rem;">${lines}</p>`;
  }

  /**
   * Send an email via the local email server
   */
  async send(options: EmailOptions): Promise<EmailResult> {
    if (!this.isConfigured) {
      console.warn('Email server not configured. Email will be logged to console.');
      console.log('📧 [Mock Email]', {
        to: options.to,
        subject: options.subject,
        text: options.text,
      });
      // Return success in development mode
      return { success: true, messageId: `mock-${Date.now()}` };
    }

    try {
      const response = await fetch(`${this.serverUrl}/api/send`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(options),
      });

      if (!response.ok) {
        throw new Error(`Email server error: ${response.statusText}`);
      }

      const result = await response.json();
      return { success: true, messageId: result.messageId };
    } catch (error: any) {
      console.error('Failed to send email:', error);
      return { success: false, error: error.message };
    }
  }

  /**
   * Send job status notification to customer
   */
  async sendJobStatusUpdate(
    customerEmail: string,
    customerName: string,
    jobId: string,
    status: string,
    vehicleInfo: string
  ): Promise<EmailResult> {
    const statusMessages: Record<string, string> = {
      'Pending': 'Your vehicle has been received and is awaiting inspection.',
      'In Progress': 'Our technicians are currently working on your vehicle.',
      'Awaiting Parts': 'We are waiting for parts to arrive. We will contact you once they are in.',
      'Awaiting Approval': 'Please review the quoted work and provide approval.',
      'Invoiced': 'Your invoice has been generated. Please review and arrange payment.',
      'Paid': 'Thank you for your payment. Your vehicle is being prepared for collection.',
      'Completed': 'Your vehicle is ready for collection. Please contact us to arrange pickup.',
      'Cancelled': 'Your job has been cancelled. Please contact us if you have any questions.',
    };

    return this.send({
      to: customerEmail,
      subject: `Job Update - ${jobId} - ${vehicleInfo}`,
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          ${this.getEmailHeader()}
          <p>Dear ${customerName},</p>
          <p>${statusMessages[status] || 'There has been an update to your job.'}</p>
          
          <div style="background: #f3f4f6; padding: 16px; border-radius: 8px; margin: 16px 0;">
            <p><strong>Job ID:</strong> ${jobId}</p>
            <p><strong>Vehicle:</strong> ${vehicleInfo}</p>
            <p><strong>Status:</strong> ${status}</p>
          </div>
          
          <p>You can log in to your customer portal to view more details about your job.</p>
          
          ${this.getEmailSignature()}
        </div>
      `,
    });
  }

  /**
   * Send invoice/quote to customer
   */
  async sendInvoice(
    customerEmail: string,
    customerName: string,
    invoiceNumber: string,
    amount: number,
    dueDate: string,
    isQuote: boolean = false
  ): Promise<EmailResult> {
    const docType = isQuote ? 'Quote' : 'Invoice';
    
    return this.send({
      to: customerEmail,
      subject: `${docType} #${invoiceNumber} - ${isQuote ? 'Request for Approval' : 'Payment Required'}`,
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          ${this.getEmailHeader()}
          <p>Dear ${customerName},</p>
          <p>Please find attached your ${docType.toLowerCase()} for services rendered.</p>
          
          <div style="background: #f3f4f6; padding: 16px; border-radius: 8px; margin: 16px 0;">
            <p><strong>${docType} Number:</strong> ${invoiceNumber}</p>
            <p><strong>Amount:</strong> R${amount.toFixed(2)}</p>
            <p><strong>${isQuote ? 'Valid Until' : 'Due Date'}:</strong> ${new Date(dueDate).toLocaleDateString('en-ZA')}</p>
          </div>
          
          ${!isQuote ? `
          <p>To pay online, please click the payment link in your customer portal or contact us.</p>
          ` : `
          <p>To approve this quote, please log in to your customer portal or contact us.</p>
          `}
          
          ${this.getEmailSignature()}
        </div>
      `,
    });
  }

  /**
   * Send appointment reminder
   */
  async sendAppointmentReminder(
    customerEmail: string,
    customerName: string,
    appointmentDate: string,
    serviceType: string,
    vehicleInfo: string
  ): Promise<EmailResult> {
    return this.send({
      to: customerEmail,
      subject: `Appointment Reminder - ${serviceType}`,
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          ${this.getEmailHeader()}
          <p>Dear ${customerName},</p>
          <p>This is a friendly reminder about your upcoming appointment.</p>
          
          <div style="background: #f3f4f6; padding: 16px; border-radius: 8px; margin: 16px 0;">
            <p><strong>Date:</strong> ${new Date(appointmentDate).toLocaleString('en-ZA')}</p>
            <p><strong>Service:</strong> ${serviceType}</p>
            <p><strong>Vehicle:</strong> ${vehicleInfo}</p>
          </div>
          
          <p>Please ensure your vehicle is available at the scheduled time.</p>
          <p>If you need to reschedule, please contact us at least 24 hours in advance.</p>
          
          ${this.getEmailSignature()}
        </div>
      `,
    });
  }

  /**
   * Send payment confirmation
   */
  async sendPaymentConfirmation(
    customerEmail: string,
    customerName: string,
    invoiceNumber: string,
    amount: number,
    paymentMethod: string
  ): Promise<EmailResult> {
    return this.send({
      to: customerEmail,
      subject: `Payment Confirmed - Invoice #${invoiceNumber}`,
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          ${this.getEmailHeader()}
          <p>Dear ${customerName},</p>
          <p>Thank you for your payment. Here are your payment details:</p>
          
          <div style="background: #f3f4f6; padding: 16px; border-radius: 8px; margin: 16px 0;">
            <p><strong>Invoice Number:</strong> ${invoiceNumber}</p>
            <p><strong>Amount Paid:</strong> R${amount.toFixed(2)}</p>
            <p><strong>Payment Method:</strong> ${paymentMethod}</p>
            <p><strong>Date:</strong> ${new Date().toLocaleDateString('en-ZA')}</p>
          </div>
          
          <p>Your vehicle is being prepared for collection. We will contact you once it is ready.</p>
          
          ${this.getEmailSignature()}
        </div>
      `,
    });
  }

  /**
   * Send welcome email to new customers
   */
  async sendWelcomeEmail(
    customerEmail: string,
    customerName: string
  ): Promise<EmailResult> {
    return this.send({
      to: customerEmail,
      subject: 'Welcome to Our Workshop Management System',
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          <h2 style="color: #2563eb;">Welcome to ${this.company.name}!</h2>
          <p>Dear ${customerName},</p>
          <p>Thank you for registering with us. You can now:</p>
          
          <ul>
            <li>Track your vehicle's service history</li>
            <li>View and pay your invoices online</li>
            <li>Schedule appointments</li>
            <li>Receive real-time job status updates</li>
          </ul>
          
          <p>If you have any questions, please don't hesitate to contact us.</p>
          
          ${this.getEmailSignature()}
        </div>
      `,
    });
  }

  /**
   * Check if email service is configured
   */
  isEmailConfigured(): boolean {
    return this.isConfigured;
  }
}

export const emailService = new EmailService();

File: backend/services/email_service.py
"""
Backend Email Service - Integration with Node.js Email Server

Sends emails for backend events like payment confirmations,
order status updates, etc. Integrates with the frontend email server.
"""

import httpx
import logging
import os
from typing import Optional, Dict, Any
from datetime import datetime

logger = logging.getLogger(__name__)


class EmailService:
    """Service for sending emails via the Node.js email server."""
    
    def __init__(self):
        self.server_url = os.getenv('EMAIL_SERVER_URL', 'http://localhost:3001')
        self.company_info = {
            'name': 'Workshop Management System',
            'email': os.getenv('COMPANY_EMAIL', 'info@workshop.co.za'),
            'phone': os.getenv('COMPANY_PHONE', ''),
            'address': os.getenv('COMPANY_ADDRESS', ''),
        }
    
    def _get_email_header(self) -> str:
        """Get HTML email header with company branding."""
        return f'<h2 style="color: #2563eb;">{self.company_info["name"]}</h2>'
    
    def _get_email_signature(self) -> str:
        """Get HTML email footer with company signature."""
        parts = [
            'Best regards,',
            f'<strong>{self.company_info["name"]}</strong>',
        ]
        
        if self.company_info['phone']:
            parts.append(f'Tel: {self.company_info["phone"]}')
        
        if self.company_info['email']:
            parts.append(f'Email: <a href="mailto:{self.company_info["email"]}">{self.company_info["email"]}</a>')
        
        if self.company_info['address']:
            parts.append(self.company_info['address'])
        
        lines = '<br>'.join(parts)
        return f'<p style="margin-top: 24px; border-top: 1px solid #e5e7eb; padding-top: 16px; color: #6b7280; font-size: 0.875rem;">{lines}</p>'
    
    async def send_payment_confirmation(
        self,
        customer_email: str,
        customer_name: str,
        invoice_number: str,
        amount: float,
        payment_method: str = 'PayFast',
        job_id: Optional[str] = None
    ) -> bool:
        """
        Send payment confirmation email to customer.
        
        Args:
            customer_email: Customer's email address
            customer_name: Customer's name
            invoice_number: Invoice number for reference
            amount: Payment amount in ZAR
            payment_method: Payment method used (default: PayFast)
            job_id: Optional job ID for context
        
        Returns:
            True if email sent successfully, False otherwise
        """
        try:
            html_content = f"""
            <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
                {self._get_email_header()}
                <p>Dear {customer_name},</p>
                <p>Thank you for your payment. Your payment has been confirmed and processed. Here are the details:</p>
                
                <div style="background: #f3f4f6; padding: 16px; border-radius: 8px; margin: 16px 0;">
                    <p><strong>Invoice Number:</strong> {invoice_number}</p>
                    <p><strong>Amount Paid:</strong> R{amount:.2f}</p>
                    <p><strong>Payment Method:</strong> {payment_method}</p>
                    <p><strong>Date:</strong> {datetime.utcnow().strftime('%d %B %Y')}</p>
                    {f'<p><strong>Job ID:</strong> {job_id}</p>' if job_id else ''}
                </div>
                
                <p>Your vehicle is being prepared for collection. We will contact you once it is ready.</p>
                <p>If you have any questions about this payment, please don't hesitate to contact us.</p>
                
                {self._get_email_signature()}
            </div>
            """
            
            return await self._send_email(
                to=customer_email,
                subject=f'Payment Confirmed - Invoice #{invoice_number}',
                html=html_content
            )
        except Exception as e:
            logger.error(f"Failed to send payment confirmation email: {e}")
            return False
    
    async def send_payment_failed(
        self,
        customer_email: str,
        customer_name: str,
        invoice_number: str,
        amount: float,
        reason: str
    ) -> bool:
        """
        Send payment failure notification to customer.
        
        Args:
            customer_email: Customer's email address
            customer_name: Customer's name
            invoice_number: Invoice number for reference
            amount: Payment amount in ZAR
            reason: Reason for payment failure
        
        Returns:
            True if email sent successfully, False otherwise
        """
        try:
            html_content = f"""
            <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
                {self._get_email_header()}
                <p>Dear {customer_name},</p>
                <p>Unfortunately, your recent payment attempt was unsuccessful. Please find the details below:</p>
                
                <div style="background: #fef2f2; border-left: 4px solid #dc2626; padding: 16px; border-radius: 8px; margin: 16px 0;">
                    <p><strong>Invoice Number:</strong> {invoice_number}</p>
                    <p><strong>Amount:</strong> R{amount:.2f}</p>
                    <p><strong>Reason:</strong> {reason}</p>
                    <p><strong>Date:</strong> {datetime.utcnow().strftime('%d %B %Y')}</p>
                </div>
                
                <p>Please try again or contact us for assistance. We accept various payment methods, and our team is happy to help.</p>
                
                {self._get_email_signature()}
            </div>
            """
            
            return await self._send_email(
                to=customer_email,
                subject=f'Payment Failed - Invoice #{invoice_number}',
                html=html_content
            )
        except Exception as e:
            logger.error(f"Failed to send payment failure email: {e}")
            return False
    
    async def send_payment_reminder(
        self,
        customer_email: str,
        customer_name: str,
        invoice_number: str,
        amount: float,
        due_date: str
    ) -> bool:
        """
        Send payment reminder email to customer.
        
        Args:
            customer_email: Customer's email address
            customer_name: Customer's name
            invoice_number: Invoice number for reference
            amount: Payment amount in ZAR
            due_date: Due date for payment
        
        Returns:
            True if email sent successfully, False otherwise
        """
        try:
            html_content = f"""
            <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
                {self._get_email_header()}
                <p>Dear {customer_name},</p>
                <p>This is a friendly reminder about your outstanding invoice payment. Payment is now due.</p>
                
                <div style="background: #fef3c7; border-left: 4px solid #f59e0b; padding: 16px; border-radius: 8px; margin: 16px 0;">
                    <p><strong>Invoice Number:</strong> {invoice_number}</p>
                    <p><strong>Amount Due:</strong> R{amount:.2f}</p>
                    <p><strong>Due Date:</strong> {due_date}</p>
                </div>
                
                <p>Please arrange payment at your earliest convenience. You can pay online through our customer portal or contact us for payment arrangements.</p>
                
                {self._get_email_signature()}
            </div>
            """
            
            return await self._send_email(
                to=customer_email,
                subject=f'Payment Reminder - Invoice #{invoice_number}',
                html=html_content
            )
        except Exception as e:
            logger.error(f"Failed to send payment reminder email: {e}")
            return False
    
    async def _send_email(self, to: str, subject: str, html: str) -> bool:
        """
        Send email via the Node.js email server.
        
        Args:
            to: Recipient email address
            subject: Email subject
            html: HTML email content
        
        Returns:
            True if email sent successfully, False otherwise
        """
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.post(
                    f"{self.server_url}/api/send",
                    json={
                        "to": to,
                        "subject": subject,
                        "html": html,
                        "text": f"Subject: {subject}"
                    }
                )
                
                if response.status_code == 200:
                    logger.info(f"Email sent successfully to {to}")
                    return True
                else:
                    logger.error(f"Email server returned {response.status_code}: {response.text}")
                    return False
        except Exception as e:
            logger.error(f"Failed to send email via server: {e}")
            return False


# Create singleton instance
email_service = EmailService()
